x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/privilege_monitoring/data_sources/sync/deletion_detection/index/deletion_detection.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from '@kbn/core/server';
+import type { ErrorCause } from 'elasticsearch-8.x/lib/api/types';
+import type { MonitoringEntitySource } from '../../../../../../../../common/api/entity_analytics/monitoring/monitoring_entity_source/monitoring_entity_source.gen';
+import type { PrivilegeMonitoringDataClient } from '../../../../engine/data_client';
+import { buildFindUsersSearchBody, getAllUserNamesInAggregation } from '../queries';
+import type { AfterKey, StaleUsersAggregations } from '../../types';
+import { findStaleUsersFactory } from '../stale_users';
+import { createBulkUtilsService } from '../../../bulk';
+import { getErrorFromBulkResponse } from '../../utils';
+
+export const createIndexDeletionDetectionService = (
+  dataClient: PrivilegeMonitoringDataClient,
+  soClient: SavedObjectsClientContract
+) => {
+  const esClient = dataClient.deps.clusterClient.asCurrentUser;
+  const findStaleUsers = findStaleUsersFactory(dataClient);
+  const bulkUtilsService = createBulkUtilsService(dataClient);
+
+  const deletionDetection = async (source: MonitoringEntitySource) => {
+    const failures: ErrorCause[] = [];
+    const allIndexUserNames = await getAllUserNamesInAggregation({
+      dataClient,
+      indexPattern: source.indexPattern,
+      buildQuery: buildFindUsersSearchBody,
+      pageSize: 100,
```

Comment:
magic number, would be good to have this as a const and maybe a comment about why 100